### PR TITLE
Yangjx fix

### PR DESCRIPTION
**Pull request by Jayson-Yang**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds configurable multi-dataset batch sampling, optional focal classification loss, and train-batch dump tooling for detection training. 🚀

### 📊 Key Changes
- Added new training args in `default.yaml` and config validation for `dump_batches`, `focal_loss`, `fl_gamma`, and `fl_alpha`.
- Introduced `ProportionalBatchSampler` and related utilities to support mixed training from multiple `train` datasets using `dataset_fractions`.
- Updated detection trainer dataloader flow to build `YOLOConcatDataset`, use proportional per-source batch sampling, and handle sampler epoch updates in DDP.
- Added `batch_dump.py` and trainer hooks to optionally copy sampled training batches to disk for dataloader and sampler inspection. 📦
- Extended detection loss logic to optionally use Focal Loss for classification, while preserving class weighting support.
- Updated label/statistics handling so plots, class weights, and auto-batch work with concatenated datasets.
- Added example local dataset YAML and a `run.txt` command showing how to launch the new workflow.

### 🎯 Purpose & Impact
- Enables more controlled multi-source detection training, which can help balance large and small datasets in a single run. 🎯
- Improves flexibility for imbalanced classification targets through optional Focal Loss settings.
- Makes debugging dataset composition and sampler behavior easier with batch preview dumping. 🔍
- Expands training capability, but also adds complexity around custom dataset YAMLs, sampling behavior, and memory usage for large combined datasets.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Jayson-Yang, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you shortly. Please review the checklist below to help ensure a smooth integration, especially since this PR appears to add new detection training functionality and configuration options. ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/build.py [hunk 1]
```diff
@@ -334,27 +341,44 @@ def build_dataloader(
...
+    generator = torch.Generator()
+    generator.manual_seed(6148914691236517205 + RANK)
+    collate_fn = getattr(dataset, "collate_fn", None)
+    prefetch_factor = 4 if nw > 0 else None
+    pin_memory = nd > 0 and pin_memory
+
+    if batch_sampler is not None:
+        # Lower prefetch to reduce RAM use when workers decode large images in parallel
+        prefetch_factor = 2 if nw > 0 else None
+        return InfiniteDataLoader(
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py:355]: 💡 **MEDIUM**: The `batch_sampler` branch no longer passes the seeded `generator` that the normal dataloader path uses. As soon as `dataset_fractions` is enabled, worker RNG initialization stops being reproducible, so identical runs can produce different augmentations/samples. Please forward `generator=generator` here as well.

File: ultralytics/data/sampler.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        n = len(d)
+        pools.append(list(range(offset, offset + n)))
+        offset += n
+    return pools
+
+
+def normalize_fractions(fractions: list[float]) -> list[float]:
+    """Normalize dataset fractions to sum to 1.0."""
+    total = float(sum(fractions))
+    if total <= 0:
```
Comment (UltralyticsAssistant) [inline ultralytics/data/sampler.py:67]: 💡 **MEDIUM**: `normalize_fractions()` only validates the sum, so values like `[-0.2, 1.2]` or other non-finite inputs can slip through and later fail inside batch construction with a much less obvious error. Please reject negative/invalid `dataset_fractions` up front before normalizing them.

File: ultralytics/cfg/datasets/coco8-multitaskdet2.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+path: D:\Lenovo_Work\Solution_Technology_Team\Dataset\component # dataset root dir
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/datasets/coco8-multitaskdet2.yaml:1]: 📝 **LOW**: This committed dataset config hardcodes a developer-local Windows path, which makes the example unusable for everyone else and brittle on non-Windows environments. Please switch this to a portable repo-relative/example path or keep it out of the tracked config set.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Useful feature set overall: the proportional sampler, concat-dataset label handling, and optional focal loss wiring look coherent. The main issues I found are a reproducibility regression in the custom `batch_sampler` dataloader path, missing validation for invalid `dataset_fractions`, and a non-portable dataset example committed with a machine-local path.

💬 Posted 3 inline comments

**Comment by glenn-jocher** [comment]
Thanks for the contribution. This PR bundles several unrelated changes under one title — a `ProportionalBatchSampler` for multi-dataset sampling, optional focal loss for the detect head, and batch-dump debug tooling — and it also includes a few changes we can't take as-is: `cfg/models/11/yolo11-cls.yaml` flips the shipped default `nc: 1000` → `nc: 2`, and `.gitignore` stops ignoring `/datasets`. We keep PRs small, focused, and single-purpose, so I'm going to close this one.

The `ProportionalBatchSampler` is the most interesting idea here — if you'd like to pursue it, please open a single focused PR (the sampler plus minimal `build.py`/trainer wiring, a small test, and a portable example dataset YAML) and we'll be happy to review. Thanks again!
